bob-common/src/mapper.rs

```diff
@@ -132,8 +133,21 @@ impl Virtual {
                     None
                 }
             })
-            .take(count)
-            .collect()
+            .collect();
+        let mut result = vec![];
+        for &node in nodes.iter() {
+            if result.len() < count && node.connection_available() {
+                result.push(node);
+            }
+        }
+        if result.len() < count {
+            for node in nodes.iter() {
+                if result.len() < count && !result.contains(&node) {
```

Comment:
How does `contains` work for `Node` struct? Does it compare the full struct content? If this is the true, then it is inefficient. It is better to look for the presence of a node by comparing only its `index`